File: ai_clients.go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/sashabaranov/go-openai"
)

func initOpenAIClient() *openai.Client {
	openAIKey := readConfig(OPENAI_KEY)
	if openAIKey == "" {
		fmt.Println("OpenAI key not found. Please set it using the config command.")
		os.Exit(1)
	}
	return openai.NewClient(openAIKey)
}

func initAnthropicClient() *Client {
	anthropicKey := readConfig(ANTHROPIC_KEY)
	if anthropicKey == "" {
		fmt.Println("Anthropic key not found. Please set it using the config command.")
		os.Exit(1)
	}
	anthropicClient, err := NewAnthropicClient(anthropicKey)
	if err != nil {
		fmt.Println("Error creating Anthropic client:", err)
		os.Exit(1)
	}
	return anthropicClient
}

func determineClient(model string) (interface{}, error) {
	openaiClient := initOpenAIClient()
	anthropicClient := initAnthropicClient()

	openaiModels, err := openaiClient.ListModels(context.Background())
	if err != nil {
		return nil, fmt.Errorf("error listing OpenAI models: %v", err)
	}

	for _, openaiModel := range openaiModels.Models {
		if openaiModel.ID == model {
			return openaiClient, nil
		}
	}

	anthropicModels := anthropicClient.ListModels()
	for _, anthropicModel := range anthropicModels {
		if anthropicModel.Name == model || anthropicModel.ApiName == model {
			return anthropicClient, nil
		}
	}

	return nil, fmt.Errorf("model %s not found in either OpenAI or Anthropic", model)
}

func listModels() {
	openaiClient := initOpenAIClient()
	anthropicClient := initAnthropicClient()

	fmt.Println("OpenAI Models:")
	openaiModels, err := openaiClient.ListModels(context.Background())
	if err != nil {
		logger.Error("Error listing OpenAI models", "error", err)
		fmt.Printf("Error listing OpenAI models: %v\n", err)
	} else {
		for _, model := range openaiModels.Models {
			fmt.Printf("- %s\n", model.ID)
		}
	}

	fmt.Println("\nAnthropic Models:")
	anthropicModels := anthropicClient.ListModels()
	for _, model := range anthropicModels {
		fmt.Printf("- %s (API Name: %s)\n", model.Name, model.ApiName)
	}
}

File: aiClients.ts
import OpenAI from "openai"
import Anthropic from "npm:@anthropic-ai/sdk"
import {ANTHROPIC_KEY, OPENAI_KEY, readConfig} from "./config.ts"
import {Model} from "openai/resources/models.ts"


let openAiClient: OpenAI | null = null
let anthropicClient: Anthropic | null = null

function initClients() {


    if (!openAiClient || !anthropicClient) {

        const openAiKey = readConfig(OPENAI_KEY)
        const anthropicKey = readConfig(ANTHROPIC_KEY)

        if (!openAiKey || !anthropicKey) {
            throw new Error("API keys are missing or invalid")
        }

        if (globalThis.isVariables) {
            console.log({openai: maskKey(openAiKey), anthropic: maskKey(anthropicKey)})
        }
        openAiClient = new OpenAI({
            apiKey: openAiKey
        })

        anthropicClient = new Anthropic({
            apiKey: anthropicKey
        })
    } else {

        if (globalThis.isVerbose) {
            console.log("Clients already initialized")
        }
    }
}

const anthropicModels = [{
    ApiName: "claude-3-5-sonnet-20240620",
    Name: "claude-3.5-sonnet",
},
{
    ApiName: "claude-3-opus-20240229",
    Name: "claude-3-opus",
},
{
    ApiName: "claude-3-sonnet-20240229",
    Name: "claude-3-sonnet",
},
{
    ApiName: "claude-3-haiku-20240307",
    Name: "claude-3-haiku",
},
]




const modelToClientMap: Record<string, OpenAI | Anthropic> = {}

export async function listModels() {
    initClients()
    const openAiModels = await initModelToClientMap()
    return {
        openai: openAiModels,
        anthropic: anthropicModels
    }
}

async function initModelToClientMap() {
    const openAiModels = await openAiClient.models.list()
    openAiModels.data.forEach((model: Model) => {
        modelToClientMap[model.id] = openAiClient
    })
    anthropicModels.forEach((model: {ApiName: string; Name: string}) => {
        modelToClientMap[model.Name] = anthropicClient
        modelToClientMap[model.ApiName] = anthropicClient
    })
    return openAiModels
}


export async function determineClient(modelName: string): {modelName: string, client: OpenAI | Anthropic} {
    initClients()
    if (!Object.keys(modelToClientMap).length) {
        await initModelToClientMap()
    }
    const client = modelToClientMap[modelName]
    if (client instanceof OpenAI) {
        return {modelName, client}
    } else {
        const anthropicModel = anthropicModels.find(model => model.Name === modelName)
        if (anthropicModel) {
            return {modelName: anthropicModel.ApiName, client}
        } else {
            throw new Error(`Model ${modelName} not found in Anthropic models`)
        }
    }

}

function maskKey(key: string): string {
    if (key.length <= 8) {
        return '*'.repeat(key.length)
    }
    return key.slice(0, 4) + '*'.repeat(key.length - 8) + key.slice(-4)
}
